### ISSUE TITLE
Lazy loading from plugins

### ISSUE THREAD

**Issue by KleinMaximus** [vueneue/vueneue#34]
I want to split the app into separate plugins, but when using lazy loading I get an error in the console.

```
import { HOMEPAGE } from './names';
import { Context } from './types';

const homepage = () => import('./views/homepage' /* webpackChunkName: "main-homepage" */);

const routes = [
  {
    path: process.env.PLUGIN_MAIN_ROUTE_HOMEPAGE || '/',
    component: homepage,
    name: HOMEPAGE,
  },
];

export default async ({ router }: Context) => {
  router.addRoutes(routes);
};
```

If I disable lazy loading

```
import homepage from './views/homepage';
```

or add a delay

```
export default async ({ router }: Context) => {
  router.addRoutes(routes);
  return new Promise(r => setTimeout(r, 0));
};
```

to the return, the error disappears.

**Comment by camwhite** [comment]
The file `src/router.js` expects an exported function that returns a `new Router()` with the routes passed into the constructor.

Here is an example https://github.com/vueneue/examples/blob/master/src/router.js

**Comment by KleinMaximus** [comment]
The documentation contains an example of adding a route inside the plugin.
https://vueneue.github.io/docs/guide/ssr.html#plugins-system

But there is an error in the console when using lazy loading:
`vue.runtime.esm.js?2b0e:587 [Vue warn]: The client-side rendered virtual DOM tree is not matching server-rendered content. This is likely caused by incorrect HTML markup, for example nesting block-level elements inside <p>, or missing <tbody>. Bailing hydration and performing full client-side render.`

**Comment by chymz** [comment]
I'm on it !

**Comment by KleinMaximus** [comment]
It is enough to do the check only on the client:
```
export default async ({ router }: Context) => {
  router.addRoutes(routes);
  return process.client ? new Promise(r => setTimeout(r, 0)) : Promise.resolve();
};
```

**Comment by chymz** [comment]
Very strange... I've done some tests:

In `src/main.js`:

```js
export default ({ router, store }) => {

  // This works without error
  router.addRoutes([
    {
      path: '/test-ok',
      component: () => import('./views/PluginRoute.vue')
    }
  ]);

  const app = new Vue({
    router,
    store,
    render: h => h(App)
  });

  // This one trigger the error
  router.addRoutes([
    {
      path: '/test-fail',
      component: () => import('./views/PluginRoute.vue')
    }
  ]);

  return app;
};
```

Actually looking for a solution...

**Comment by KleinMaximus** [comment]
@chymz 
Will there be an error if move the second `addRoutes` before app initializing?

**Comment by chymz** [comment]
Nope.

I've investigated a bit on Nuxt codebase, and it seems that plugins initializations are called before `new Vue`

So I'm doing wrong actually with VueNeue, but the fix will introduce a big breaking change: `context.app` will not be available in plugin init function.

**Comment by KleinMaximus** [comment]
I think it is possible to create some fake context object for the  plugins initialization time, which can then be injected to the application after full complete plugins settings.

**Comment by chymz** [comment]
Yup I think I will replace `app` variable with a `appCreated` function for plugins:

```js
export default ({ appCreated }) => {
  appCreated(app => {
    // ...
  });
}
```

I will work on this tomorrow !

**Comment by chymz** [comment]
Fixed in `0.3.20`
